File: crackup/lib/CrackupDriver.rb
require 'uri'

module Crackup

  # Base storage driver class for Crackup.
  class CrackupDriver
    attr_reader :url
    
    # Gets an instance of the appropriate storage driver to handle the specified
    # <em>url</em>. If no suitable driver is found, raises an error of type
    # StandardError.
    def self.get_driver(url)
      begin
        uri = URI::parse(url)
      rescue => e
        abort("Invalid URL: #{url}")
      end
      
      # Use the filesystem driver if no scheme is specified or if the scheme is
      # a single letter (which indicates a Windows drive letter).
      if uri.scheme.nil? || uri.scheme =~ /^[a-z]$/i
        scheme = 'File'
      else
        scheme = uri.scheme.capitalize
      end
      
      driver_class = "CrackupDriver#{scheme}"
      driver_file  = File.dirname(__FILE__) + "/../drivers/#{driver_class}.rb"
      
      # Load the driver.
      unless require(driver_file)
        raise StandardError, "Driver not found: #{uri.scheme}"
      end
      
      return Crackup::const_get(driver_class).new(url)
    end
    
    def initialize(url)
      @url = url
    end
    
    def delete(url)
      return false
    end
    
    def get(url, local_filename)
      return false
    end
    
    def put(url, local_filename)
      return false
    end
  end
end

File: crackup-php/classes/CrackupDriver.php
<?php
/**
 * CrackupDriver is a base class representing a Crackup storage driver.
 * 
 * @author Ryan Grove <ryan@wonko.com>
 * @package Crackup
 */
abstract class CrackupDriver {
  // -- Private Instance Variables ---------------------------------------------
  private $_url;
  
  // -- Public Static Methods --------------------------------------------------

  /**
   * Parses a URL and returns an instance of the appropriate Crackup storage
   * driver (if any).
   * 
   * @param string $url
   * @return CrackupDriver
   */
  public static function getDriver($url) {
    // Parse the URL.
    if (false === ($parsedUrl = @parse_url($url))) {
      throw new CrackupDriverUrlException('Invalid URL: '.$url);
    }
    
    // Look for the driver.
    $driverName  = (isset($parsedUrl['scheme']) ? 
        strtolower($parsedUrl['scheme']) : 'file');
    $driverClass = 'CrackupDriver'.ucfirst($driverName);
    $driverFile  = dirname(__FILE__).'/../drivers/'.$driverClass.'.php';
    
    if (!file_exists($driverFile)) {
      throw new CrackupDriverNotFoundException('Driver not found: '.
          $driverName);
    }
    
    // Load the driver.
    require_once $driverFile;
    
    return new $driverClass($url);
  }
  
  // -- Public Instance Methods ------------------------------------------------
  public function __construct($url) {
    $this->_url = $url;
  }
  
  // -- Abstract Public Instance Methods ---------------------------------------  
  abstract public function delete($filename);  
  abstract public function get($remote_filename, $local_filename);
  abstract public function put($local_filename, $remote_filename);
}

class CrackupDriverException extends Exception {}
class CrackupDriverDeleteException extends CrackupDriverException {}
class CrackupDriverGetException extends CrackupDriverException {}
class CrackupDriverNotFoundException extends CrackupDriverException {}
class CrackupDriverPutException extends CrackupDriverException {}
class CrackupDriverUrlException extends CrackupDriverException {}
?>